src/rdkafka_txnmgr.c

```diff
@@ -2027,21 +2034,37 @@ rd_kafka_error_t *rd_kafka_send_offsets_to_transaction(
  *
  * Current state must be either COMMIT_NOT_ACKED or ABORT_NOT_ACKED.
  *
+ * @returns true if the transaction was completed, false if an epoch bump
+ *          is needed before completing.
+ *
  * @locality rdkafka main thread
  * @locks rd_kafka_wrlock(rk) MUST be held
  */
-static void rd_kafka_txn_complete(rd_kafka_t *rk, rd_bool_t is_commit) {
+static rd_bool_t rd_kafka_txn_complete(rd_kafka_t *rk, rd_bool_t is_commit) {
         rd_kafka_dbg(rk, EOS, "TXNCOMPLETE", "Transaction successfully %s",
                      is_commit ? "committed" : "aborted");
 
         /* Clear all transaction partition state */
         rd_kafka_txn_clear_pending_partitions(rk);
         rd_kafka_txn_clear_partitions(rk);
 
+        rd_bool_t drain_bump =
+            rk->rk_eos.txn_requires_epoch_bump ||
+            rk->rk_eos.idemp_state != RD_KAFKA_IDEMP_STATE_ASSIGNED;
+
         rk->rk_eos.txn_requires_epoch_bump = rd_false;
         rk->rk_eos.txn_req_cnt             = 0;
 
-        rd_kafka_txn_set_state(rk, RD_KAFKA_TXN_STATE_READY);
+        if (drain_bump) {
```

Comment:
This function, txn_complete(), should be called when the entire transaction is done, including bumps, so this drain_bump logic should not be here but in an earlier stage.
Recommend adding a `.._try_complete()` that contains this logic.